csrc/scheduler/normalization_utils.cpp

```diff
@@ -698,11 +702,61 @@ std::optional<int64_t> getOptionalInnerOuterPersistentBufferBatches(
       inner_batch,
       ", batch_min = ",
       batch_min);
+
   const int64_t batch_max = getMaximumInnerOuterPersistentBufferBatch();
-  if (inner_batch <= batch_max) {
-    return inner_batch;
+
+  // consider non-divisible splits.
+  // If inner_bath is still larger than bath_max and threads_per_block is
+  // smaller than threads_per_block_max possible. Try non-divisible batch
+  // splits.
+  while (inner_batch > batch_max &&
+         threads_per_block * 2 <= threads_per_block_max) {
+    threads_per_block *= 2;
+    inner_batch = ceilDiv(after_vectorization, threads_per_block);
+  }
+
+  // If inner_batch is still larger than batch_max, try set threads_per_block to
+  // max possible and recheck inner_batch, e.g. Initial after_vectorization =
+  // 9280/8 = 1160, threads_per_block = 8, inner_batch = 1160/8 = 145. After
+  // dividible split,  inner_batch = 29 and threads_per_block = 40. After
+  // non-divisible split, inner_batch = 8 and threads_per_block = 160. The
+  // following if statement will set threads_per_block to 256 and inner_batch
+  // to 5.
+  if (inner_batch > batch_max) {
+    threads_per_block = threads_per_block_max;
+    inner_batch = ceilDiv(after_vectorization, threads_per_block);
+  }
+
+  // If threads_per_block is smaller than threads_per_block_min, move
+  // parallelism from inner_batch to threads_per_block. e.g. Initial
+  // after_vectorization = 1344/8 = 168, threads_per_block = 8, inner_batch
+  // = 21. Up to there, threads_per_block = 24, inner_batch = 7. The following
+  // code will set threads_per_block to 128 and inner_batch to 2. And the
+  // latency for batch size 16K is reduced from 295 us to 150 us.
+  if (threads_per_block < threads_per_block_min) {
+    threads_per_block = threads_per_block_min;
+    inner_batch = ceilDiv(after_vectorization, threads_per_block);
+  }
+
+  // The maximum feature size can be processed without register spills and
+  // fusion segmentation for fp16 is 14K. Here, we can allow register spills to
+  // avoid fusion segmentation by incrase maximum batch size by 3. This allows
+  // us to process up to 20K features (14K + 256*8*3).
+  // Performance on A100-80G:
+  // (1) shape= 16384 x 16384, 1300 GB/s, time_us mean(var)= 1245.08 (8.89703),
+  // 64 bytes stack frame, 64 bytes spill stores, 128 bytes spill loads. (2)
+  // shape= 16384 x 18432, 1070 GB/s, time_us mean(var)= 1683.87 (19.527), 192
+  // bytes stack frame, 192 bytes spill stores, 384 bytes spill loads.
+  // (3) shape= 16384 x 20480, 730 GB/s time_us mean(var)= 2766.64 (12.3883),
+  // 320 bytes stack frame, 320 bytes spill stores, 640 bytes spill loads. As a
+  // ref, the segmented version takes time_us mean(var)= 2841.91 (5.20231)
+  // without considering the overhead of fusion segmentation.
+  const int64_t batch_max_reg_spill = batch_max + 3;
+
+  if (enforce_return_valid || inner_batch <= batch_max_reg_spill) {
```

Comment:
So this means we accept whatever batch sizes as a valid heuristics parameter, no matter how much register usage blows up?

Reply:
> So this means we accept whatever batch sizes as a valid heuristics parameter, no matter how much register usage blows up?

Yes. We already have a runtime check of buffer size before this function. So register spill is controlled.